default.nix

```diff
@@ -138,12 +62,21 @@ in {
       runHook postInstall
     '';
 
-    installCheckPhase = ''
-      runHook preInstallCheck
+    # TODO: It would be a great idea to run epubcheck and succeed, but since we moved to mdBook, the plugin that generates an epub creates an output that fails epubcheck in many ways that we can't fix ourselves.
+    # In any case, the epub was manually inspected and has all the content displayed without big issues.
```

Comment:
If some viewer does display it, it does not mean others will be able to as well (they might have stricter parsers). Looks like there are two kinds of issues:

- In `OEBPS/content.opf` file, the `id` and `idref` attributes start with a number (since `mdbook-epub` derives them from a file name). Maybe we could work around it by using `pXX` prefix on files, or in the `content.opf`
- The interfile md links are not updated in text bodies.

The following workaround appears to work:

```patch
diff --git a/anchors.jq b/anchors.jq
new file mode 100644
index 0000000..7230977
--- /dev/null
+++ b/anchors.jq
@@ -0,0 +1,31 @@
+"\\[\\]\\{#(?<anchor>[^\\}]+?)\\}" as $empty_anchor_regex |
+"\\[(?<text>[^\\]]+?)\\]\\{#(?<anchor>[^\\}]+?)\\}" as $anchor_regex |
+
+
+def transform_anchors_html:
+    . | gsub($empty_anchor_regex; "<a name=\"" + .anchor + "\"></a>")
+      | gsub($anchor_regex; "<a href=\"#" + .anchor + "\" id=\"" + .anchor + "\">" + .text + "</a>");
+
+
+def transform_anchors_strip:
+    . | gsub($empty_anchor_regex; "")
+      | gsub($anchor_regex; .text);
+
+
+def map_contents_recursively(transformer):
+    . + {
+        Chapter: (.Chapter + {
+            content: .Chapter.content | transformer,
+            sub_items: .Chapter.sub_items | map(map_contents_recursively(transformer)),
+        }),
+    };
+
+
+def process_command:
+    .[0] as $context |
+    .[1] as $body |
+    $body + {
+        sections: $body.sections | map(map_contents_recursively(if $context.renderer == "html" then transform_anchors_html else transform_anchors_strip end)),
+    };
+
+process_command
diff --git a/book.toml b/book.toml
index 2e4902e..affaa23 100644
--- a/book.toml
+++ b/book.toml
@@ -33,6 +33,10 @@ git-repository-url = "https://github.com/NixOS/nix-pills"
 "fundamentals-of-stdenv.html" = "19-fundamentals-of-stdenv.html"
 "basic-dependencies-and-hooks.html" = "20-basic-dependencies-and-hooks.html"
 
+[preprocessor.anchors]
+renderers = ["html"]
+command = "jq --from-file anchors.jq"
+
 [output.linkcheck]
 follow-web-links = true
 
diff --git a/default.nix b/default.nix
index ae3ed31..1bee8f9 100644
--- a/default.nix
+++ b/default.nix
@@ -1,17 +1,15 @@
 { pkgs ? import <nixpkgs> {} }:
-let
-  buildInputs = with pkgs; [
-    mdbook
-    mdbook-epub
-    mdbook-linkcheck
-  ];
-in
+
 {
   html-split = pkgs.stdenvNoCC.mkDerivation {
     name = "nix-pills";
     src = ./.;
 
-    inherit buildInputs;
+    nativeBuildInputs = with pkgs; [
+      jq
+      mdbook
+      mdbook-linkcheck
+    ];
 
     buildPhase = ''
       # We can't check external links inside the sandbox, but it's good to check them outside the sandbox.
@@ -39,10 +37,29 @@ in
     name = "nix-pills-epub";
     src = ./.;
 
-    inherit buildInputs;
+    nativeBuildInputs = with pkgs; [
+      jq
+      mdbook-epub
+      unzip
+      zip
+    ];
 
     buildPhase = ''
       mdbook-epub --standalone .
+
+      mkdir nix-pills.epub-fix
+      ( cd nix-pills.epub-fix
+        unzip -q "../book/epub/Nix Pills.epub"
+        # Fix invalid ids.
+        sed -Ei 's/(id(ref)?=")([0-9])/\1p\3/g' OEBPS/content.opf
+        sed -Ei 's/(id="|href="#)([0-9])/\1fn\2/g' OEBPS/20-basic-dependencies-and-hooks.html
+        # Fix anchors, mdbook-epub does not run preprocessors.
+        sed -Ei 's~\[\]\{#([^\}]+)\}~<a id="\1" />~g' OEBPS/*.html
+        # Fix broken links in body.
+        sed -Ei 's/("[0-9a-z-]+\.)md(["#])/\1html\2/g' OEBPS/*.html
+        zip -q "../book/epub/Nix Pills.epub" **/*
+      )
+      # rm -r nix-pills.epub-fix
     '';
 
     installPhase = ''
@@ -65,18 +82,18 @@ in
     # TODO: It would be a great idea to run epubcheck and succeed, but since we moved to mdBook, the plugin that generates an epub creates an output that fails epubcheck in many ways that we can't fix ourselves.
     # In any case, the epub was manually inspected and has all the content displayed without big issues.
     # Because of this, we leave the code below as a comment so someone in the future can try enabling it again, or working on the mdBook epub plugin to fix the issues caught by epubcheck.
-    # installCheckInputs = with pkgs; [
-    #   epubcheck
-    # ];
+    installCheckInputs = with pkgs; [
+      epubcheck
+    ];
 
-    # doInstallCheck = true;
+    doInstallCheck = true;
 
-    # installCheckPhase = ''
-    #   runHook preInstallCheck
+    installCheckPhase = ''
+      runHook preInstallCheck
 
-    #   epubcheck "$manual"
+      epubcheck "$manual"
 
-    #   runHook postInstallCheck
-    # '';
+      runHook postInstallCheck
+    '';
   };
 }
diff --git a/pills/03-enter-environment.md b/pills/03-enter-environment.md
index 66b140c..dcb185e 100644
--- a/pills/03-enter-environment.md
+++ b/pills/03-enter-environment.md
@@ -1,4 +1,4 @@
-# Enter the Environment {#enter-environment}
+# Enter the Environment []{#enter-environment}
 
 Welcome to the third Nix pill. In the [second pill](02-install-on-your-running-system.md) we installed Nix on our running system. Now we can finally play with it a little, these things also apply to NixOS users.
 
@@ -16,7 +16,7 @@ $ source ~/.nix-profile/etc/profile.d/nix.sh
 
 To remind you, `~/.nix-profile/etc` points to the `nix-2.1.3` derivation. At this point, we are in our Nix user profile.
 
-## Install something
+## Install something []{#install-something}
 
 Finally something practical! Installation into the Nix environment is an interesting process. Let's install `hello`, a simple CLI tool which prints `Hello world` and is mainly used to test compilers and package installations.
 
diff --git a/pills/20-basic-dependencies-and-hooks.md b/pills/20-basic-dependencies-and-hooks.md
index 51dcac4..85a2769 100644
--- a/pills/20-basic-dependencies-and-hooks.md
+++ b/pills/20-basic-dependencies-and-hooks.md
@@ -273,4 +273,5 @@ and if one dependency has that setup hook then all of them will be so `echo`ed.
 ...I'm not sure! We could talk about the additional dependency types and hooks which cross compilation necessitates, building on our knowledge here to cover stdenv as it works today. We could talk about how nixpkgs is bootstrapped. Or we could talk about how `localSystem` and `crossSystem` are elaborated into the `buildPlatform`, `hostPlatform`, and `targetPlatform` each bootstrapping stage receives. Let us know which most interests you!
 
 [^1]: We can now be precise and consider what `addToEnv` does alone the minimal treatment of a dependency: i.e. a package that is _just_ a dependency would _only_ have `addToEnv` applied to it.
+
 [^2]: It was called [GCC Wrapper](https://github.com/NixOS/nixpkgs/tree/6675f0a52c0962042a1000c7f20e887d0d26ae25/pkgs/build-support/gcc-wrapper) in the version of nixpkgs suggested for following along in this pill; Darwin and Clang support hadn't yet motivated the rename.
```

Reply:
@jtojnar do you want to apply this on the branch, so it will be recorded on your name, or do you prefer me to apply the patch?